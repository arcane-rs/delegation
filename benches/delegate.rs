use criterion::{criterion_group, criterion_main, Criterion};
use delegation::delegate;

#[delegate]
trait AsStr {
    fn as_str(&self) -> &str;
}

impl AsStr for String {
    fn as_str(&self) -> &str {
        self
    }
}

macro_rules! def_enum {
    ($($variant:ident),+) => {
        #[delegate(derive(AsStr))]
        enum Name {
            $($variant(String)),+
        }
    };
}

macro_rules! def_bench {
    ($($variant:ident),*) => {
        fn bench(bencher: &mut ::criterion::Bencher) {
            $(
                bencher.iter(|| {
                    let name = ::criterion::black_box(
                        Name::$variant(String::from("Test"))
                    );

                    assert_eq!(name.as_str(), "Test");
                });
            )*
        }
    };
}

macro_rules! define {
    ($($variant:ident),*) => {
        def_enum!($($variant),*);
        def_bench!($($variant),*);
    };
}

define!(
    A, B, C, D, E, F, G, H, I, J, K, L, M, N, O, P, Q, R, S, T, U, V, W, X, Y,
    Z
);

fn delegate_benchmark(c: &mut Criterion) {
    c.bench_function("delegate", bench);
}

criterion_group!(benches, delegate_benchmark);
criterion_main!(benches);
