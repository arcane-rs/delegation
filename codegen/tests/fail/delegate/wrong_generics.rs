use delegation::delegate;

#[derive(Clone, Copy, Debug, PartialEq)]
struct Ver<const V: u8>;

#[delegate]
trait Versioned<const V: u8> {
    fn version(&self) -> Ver<{ V }>;
}

struct Created;

impl<const V: u8> Versioned<V> for Created {
    fn version(&self) -> Ver<V> {
        Ver::<V>
    }
}

#[delegate(derive(
    for<'a> Versioned<'a>,
))]
enum Events {
    Create(Created),
}

fn main() {}
