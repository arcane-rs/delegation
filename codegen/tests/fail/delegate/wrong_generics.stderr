error[E0412]: cannot find type `V` in this scope
  --> tests/fail/delegate/wrong_generics.rs:7:23
   |
7  |   trait Versioned<const V: u8> {
   |                         ^ not found in this scope
...
19 | / #[delegate(derive(
20 | |     for<'a> Versioned<'a>;
21 | | ))]
   | |___- in this procedural macro expansion
   |
   = note: this error originates in the macro `Versioned` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you might be missing a type parameter
   |
20 |     for<'a, V> Versioned<'a>;
   |           +++

error[E0107]: trait takes 0 lifetime arguments but 1 lifetime argument was supplied
  --> tests/fail/delegate/wrong_generics.rs:20:13
   |
20 |     for<'a> Versioned<'a>;
   |             ^^^^^^^^^---- help: remove the unnecessary generics
   |             |
   |             expected 0 lifetime arguments
   |
note: trait defined here, with 0 lifetime parameters
  --> tests/fail/delegate/wrong_generics.rs:7:7
   |
7  | trait Versioned<const V: u8> {
   |       ^^^^^^^^^

error[E0107]: trait takes 1 generic argument but 0 generic arguments were supplied
  --> tests/fail/delegate/wrong_generics.rs:20:13
   |
20 |     for<'a> Versioned<'a>;
   |             ^^^^^^^^^ expected 1 generic argument
   |
note: trait defined here, with 1 generic parameter: `V`
  --> tests/fail/delegate/wrong_generics.rs:7:7
   |
7  | trait Versioned<const V: u8> {
   |       ^^^^^^^^^ -----------
help: add missing generic argument
   |
20 |     for<'a> Versioned<'a, V>;
   |                         +++

error[E0747]: unresolved item provided when a constant was expected
  --> tests/fail/delegate/wrong_generics.rs:7:23
   |
7  |   trait Versioned<const V: u8> {
   |                         ^
...
19 | / #[delegate(derive(
20 | |     for<'a> Versioned<'a>;
21 | | ))]
   | |___- in this procedural macro expansion
   |
   = note: this error originates in the macro `Versioned` which comes from the expansion of the attribute macro `delegate` (in Nightly builds, run with -Z macro-backtrace for more info)
help: if this generic argument was intended as a const parameter, surround it with braces
   |
7  | trait Versioned<const { V }: u8> {
   |                       +   +

error[E0107]: trait takes 1 generic argument but 0 generic arguments were supplied
  --> tests/fail/delegate/wrong_generics.rs:20:13
   |
20 |     for<'a> Versioned<'a>;
   |             ^^^^^^^^^ expected 1 generic argument
   |
note: trait defined here, with 1 generic parameter: `V`
  --> tests/fail/delegate/wrong_generics.rs:7:7
   |
7  | trait Versioned<const V: u8> {
   |       ^^^^^^^^^ -----------
help: add missing generic argument
   |
6  | #[delegate]V
   |            +
